source/templates/writing-helpers.md

```diff
@@ -1,62 +1,383 @@
+
+```hbs
+Your total is $2.50.
+```
+
+Whenever the arguments you've passed to a helper change, whether they
+come from a model or a component, Ember will automatically call your
+helper again with the new values and keep the DOM up-to-date.
+
+### Helper Names
 
-Anywhere in your Handlebars templates, you can now invoke this helper:
+Unlike components, helpers do not require a dash (`-`) character in
+their name.
+
+### Helper Arguments
+
+You can pass one or more arguments to Handlebars helpers which can be used
+inside the function. In the above example, we passed the amount in cents
+as the first and only argument.
+
+To pass multiple arguments to a helper, add them as a space-separated
+list after the helper name:
 
 ```handlebars
-{{highlight name}}
+{{my-helper "hello" "world"}}
 ```
 
-and it will output the following:
+An array of these arguments is passed to the helper function:
+
+```app/helpers/my-helper.js
+import Ember from "ember";
 
-```html
-<span class="highlight">Peter</span>
+export default Ember.Helper.helper(function(params) {
+  let arg1 = params[0];
+  let arg2 = params[1];
+
+  console.log(arg1); // => "hello"
+  console.log(arg2); // => "world"
+});
 ```
 
-If the `name` property on the current context changes, Ember.js will
-automatically execute the helper again and update the DOM with the new
-value.
+You can use JavaScript's destructuring assignment shorthand to clean up
+the code. This example is equivalent to the above example (note the function signature):
 
-### Undasherized helpers
+```app/helpers/my-helper.js
+import Ember from "ember";
 
-Helpers without a dash in their name won't be autoloaded by Ember. You can still use them, but you'll need to register them in an initializer, or in `app/app.js`.
+export default Ember.Helper.helper(function([arg1, arg2]) {
+  console.log(arg1); // => "hello"
+  console.log(arg2); // => "world"
+});
+```
+
+### Named Arguments
+
+Normal arguments are useful for passing data to be transformed into
+helper functions. However, because the order in which you pass arguments
+matters, it is usually best not to have helpers take more than one or
+two of them.
+
+That said, sometimes you may want to make behavior of helpers
+configurable by the developers that call them from their templates. For
+example, let's abandon our American-centric ways and update our
+`format-currency` helper to take an optional configuration for which
+currency symbol to display.
 
-```app/app.js
-Ember.Handlerbars.registerHelper('highlight', highlight);
+Handlebars helpers allow you to pass named arguments as a JavaScript
+object that contains the name of the argument along with an associated
+value.  The order in which named arguments are supplied does not affect
+functionality.
+
+In this example, we can pass a `sign` argument to our `format-currency`
+helper:
+
+```handlebars
+{{format-currency 350 sign="£"}}
 ```
 
-### Dependencies
+Now our helper will print pounds sterling rather than US dollars:
```

Comment:
"We'd like our helper to print pounds..." (trying not to imply that it already will somehow magically do this)